Command line arguments

kofigumbs opened the pull request:
@trptcolin @robertjlooby Feature-wise, this adds various command line options (see README.md)

kofigumbs committed:
Expressively rename variables and fields

kofigumbs committed:
Use configurations to setup Simulation

- Rename Simulation to DefaultSimulation
- Simulation is now interface implemented by DefaultSimulation
- Move main() to src/.../main/java/Main.java

kofigumbs committed:
Clean names

- Added test for SizeConfigurationTest to check exception behavior

kofigumbs committed:
Move static Rules dependencies to instance

- Rename Rules to DefaultRules which implements Rules
- Dynamically create Rules.winningCombinations based on size

kofigumbs committed:
Move away from relying on concrete Players

- Introduce PlayerConfiguration
- Extended functinoality of Player interface

kofigumbs committed:
Move wiring responsibility to Configurations (from Simulation)

- Use Configuration in Main
- Add n*n playable games (MINIMAX IS TOO SLOW)
- Extend Simulation, Player, and Rules interfaces to work with Configurations

kofigumbs committed:
Fix replay bug

kofigumbs committed:
Fix minor readability stuff

- Add brackets to empty while loop
- Welcome user before asking them if they want to go first
- Minimax prints its optimized move

trptcolin commented:
There's some weird behavior here where it asks for user input before printing out the available options:

```
colin:tictactoe-java/ (command-line-arguments) $ java -jar build/libs/tictactoe-java.jar --help
Welcome to Tic Tac Toe!

Would you like to go first? (y/n) y
Make a move by entering an empty space id
usage: java -jar tictactoe-java.jar [--size <number>] [--padding <number>] [--human|--minimax]
```

Also, this PR takes the code coverage down from 97% to 92% - let's get that back up above 95% (and to 100% if possible).

trptcolin commented on src/me/hkgumbs/tictactoe/main/java/Main.java:
```diff
@@ -0,0 +1,41 @@
+
+        try {
+            for (Configuration configuration : CONFIGURATIONS)
+                configuration.apply(arguments, simulation);
```
This design is nice, because it means you can add new configurations pretty easily.

It might be nice to separate this logic into another class, because looks like it'd be useful to test (things like catching exceptions, what to do if arguments are empty, how the configurations are applied, etc.). And `main()` is usually pretty tricky to get comprehensive tests around.

trptcolin commented on src/me/hkgumbs/tictactoe/main/java/simulation/SizeConfiguration.java:
```diff
@@ -0,0 +1,30 @@
+    private static final String KEY = "--size";
+    private static final int DEFAULT_SIZE = 3;
+
+    private int extractSize(List<String> args) {
```
This method looks pretty much identical to `extractPadding` in `FormatterConfiguration`. Surely there's a way to DRY that up - maybe a small helper class with (at least) one method that takes the argument list, the argument name, and the default value?

trptcolin commented on src/me/hkgumbs/tictactoe/test/java/simulation/RulesConfigurationTest.java:
```diff
@@ -0,0 +1,31 @@
+            for (Configuration configuration : configurations)
+                configuration.apply(Arrays.asList(), simulation);
+        } catch (Configuration.CannotApplyException e) {
+            e.printStackTrace();
```
What should happen in a test run if this exception gets thrown? Should the test fail? The way this is written, we'd see a stack trace, but no test would fail.

trptcolin commented on src/me/hkgumbs/tictactoe/test/java/simulation/SizeConfigurationsTest.java:
```diff
@@ -0,0 +1,70 @@
+        try {
+            configuration.apply(args, simulation);
+        } catch (Configuration.CannotApplyException e) {
+            e.printStackTrace();
```
Same question as before: should tests fail if an exception is caught here?

trptcolin commented on src/me/hkgumbs/tictactoe/test/java/simulation/SizeConfigurationsTest.java:
```diff
@@ -0,0 +1,70 @@
+        } catch (Configuration.CannotApplyException e) {
+            flag = true;
+        }
+        assertTrue(flag);
```
It's always tricky to interpret error messages that say things like "Expected true, got false". I'd suggest adding a descriptive message using [this arity of `assertTrue`](http://junit.sourceforge.net/javadoc/org/junit/Assert.html#assertTrue(java.lang.String, boolean)) that allows a message to be displayed.

Another thing you could do to make this more readable would be to rename `flag` to something more descriptive: *why* does this thing get set to `true`?

trptcolin commented:
Lots of good stuff in here. To name a few: some really nice renames, extracting the `Configuration` objects, and taking much better advantage of your `Player` abstraction.

kofigumbs committed:
Clean configuration tests and extract Duplication from number configurations

robertjlooby commented on src/me/hkgumbs/tictactoe/main/java/simulation/Extractor.java:
```diff
@@ -0,0 +1,16 @@
+import java.util.List;
+
+public class Extractor {
+    public static int extract(List<String> args, String key, int defaultValue)
```
`Extractor.extract` isn't a very descriptive name...

kofigumbs committed:
Rename Extractor.extract to parseInt

hkgumbs commented:
@trptcolin @robertjlooby I am concerned about the Simulation interface, since it basically just exposes the private variables in DefaultSimulation. I read in Clean Code about data structures, and just wanted to make sure it was still a good practice since it was new to me. I'm thinking of just having a Simulation class that just has the 5 or so public variables that keep track of the game, then moving the state change logic to a separate controller class. Then I can pass the Simulation around to the Players as well instead of the setter/getter exposing I'm doing there.

robertjlooby commented on src/me/hkgumbs/tictactoe/main/java/Main.java:
```diff
@@ -0,0 +1,41 @@
+    private static final String USAGE = "usage: java -jar tictactoe-java.jar" +
+            " [--size <number>]" +
+            " [--padding <number>]" +
+            " [--humans|--minimax]";
```
It would be nice to document the options in the README

hkgumbs replied:
@robertjlooby done

trptcolin commented:
@hkgumbs I agree: by providing a wide interface like `Simulation`, you make it possible for Players to do things like swap out the other player (!) - what if a Player implementation wanted to cheat and pass in a TerriblePlayer implementation? 

So I agree that this thing might have 2 jobs: wiring and playing the game.

One nice thing about the `Simulation` interface, even though it looks like it's just exposing private variables, is that it decouples the storage mechanism from how you tell the thing to get/change state. I think you can retain this benefit with the split you're suggesting in your comment.

kofigumbs committed:
Decouple Simulation storage and logic

- Change Simulation to abstract class with public variables
- Fix board not printing after game is over
- Upped test coverage to 97% (by dissembling Simulation.next)
- Fix replay bug in human v minimax
- Simplify Player interface by using Simulation as encapsulation
- Introduce StubSimulation for testing
- Introduce Controller for simulation logic
- Update README.md to list command-line options

hkgumbs commented:
@trptcolin @robertjlooby I apologize for yet another lengthy commit. As I was pulling DefaultSimulation apart, I had to refactor many times, and it seemed like any intermediate state would have left the game broken

kofigumbs committed:
Simplify DefaultController

- Move snapshot logic to its own class (Snapshot)
- Convert start recursion to while loop

kofigumbs committed:
Rename packages (change no code)

kofigumbs committed:
Remove duplication and legacy cruft

kofigumbs committed:
Implement alpha-beta pruning in Minimax

MinimaxTest imposes a thirty-second limit on the second move
of a 4x4 game. This should be decreased as improvements are made.

kofigumbs committed:
Fix clunky printing by defaulting control to first player

kofigumbs committed:
Speed up Minimax using caching

hkgumbs commented:
@trptcolin @robertjlooby 
- [X] Speed up Minimax to less than 1.5 seconds
- [X] Removed double printing/asking in human v human

hkgumbs commented on src/me/hkgumbs/tictactoe/main/java/rules/DefaultRules.java:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    @Override
+    public void printWinnerMessage(Board board) {
```
Not sure if this is the right decision here, since it forces the Rules to deal with output, but it would be duplicated if I put it in Player concretions, and it doesn't seem like a big enough idea fro its own class.

robertjlooby replied:
I agree the `Rules` shouldn't care about output. This and `generateWinnerMessage` should be moved somewhere else so the `Rules` don't care about actual output (i.e. the `Rules` class should not depend on `OutputStream` at all)

kofigumbs committed:
Remove output logic from Rules

hkgumbs commented:
@robertjlooby @trptcolin iterated on all of the feedback here. I think this and #4 finish out this project.

trptcolin commented:
@hkgumbs looks good to me. Very nice speed improvement. It took awhile for me to figure out how to play human vs. minimax (user must omit flags), so it might be nice to provide some default. I bet you've got that solved w/ the dumb computer branch though, so I'll go ahead & merge.
